src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKeySynchronizer.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class AccessKeySynchronizer : IAccessKeySynchronizer, IDisposable
+    {
+        private readonly ConcurrentDictionary<ServiceEndpoint, object> _endpoints = new ConcurrentDictionary<ServiceEndpoint, object>(ReferenceEqualityComparer.Instance);
+
+        private readonly ILoggerFactory _factory;
+        private readonly IServerNameProvider _provider;
+        private readonly TimerAwaitable _timer = new TimerAwaitable(TimeSpan.Zero, TimeSpan.FromMinutes(5));
+
+        public AccessKeySynchronizer(
+            IServerNameProvider nameProvider,
+            ILoggerFactory loggerFactory
+            )
+        {
+            _ = UpdateAsync();
+
+            _provider = nameProvider;
+            _factory = loggerFactory;
+        }
+
+        /// <summary>
+        /// For test only.
+        /// </summary>
+        private AccessKeySynchronizer()
+        {
+            _provider = new DefaultServerNameProvider();
+            _factory = NullLoggerFactory.Instance;
+        }
+
+        internal static AccessKeySynchronizer ForTest() => new AccessKeySynchronizer();
+
+        public void AddServiceEndpoint(ServiceEndpoint endpoint)
+        {
+            if (endpoint.AccessKey is AadAccessKey aadKey)
+            {
+                _ = aadKey.UpdateAccessKeyAsync(_provider, _factory);
```

Comment:
What if update failed? Still add to _endpoints?

Reply:
Yes, add to endpoints so the `accesskey` could be updated again after 5 minutes.